src/python_framework_v02/nhd_network.py

```diff
@@ -264,3 +268,57 @@ def kahn_toposort_edges(N):
     for n in sorted_nodes:
         for m in N.get(n, ()):
             yield (n, m)
+
+
+def reservoir_boundary(connections, waterbodies, n):
+    if n not in waterbodies and n in connections:
+        return any(x in waterbodies for x in connections[n])
+    return False
+
+
+def separate_waterbodies(connections, waterbodies):
+    waterbody_nodes = {}
+    for wb, nodes in reverse_surjective_mapping(waterbodies).items():
+        waterbody_nodes[wb] = net = {}
+        for n in nodes:
+            if n in connections:
+                net[n] = list(filter(waterbodies.__contains__, connections[n]))
+    return waterbody_nodes
+
+
+def replace_waterbodies_connections(connections, waterbodies):
+    """
+    Use a single node to represent waterbodies. The node id is the
+    waterbody id.
+
+    This returns a new copy of connections with transformation
+    """
+    new_conn = {}
+    waterbody_nets = separate_waterbodies(connections, waterbodies)
+    for n in connections:
+        if reservoir_boundary(connections, waterbodies, n):
+            # one of the children of n is a member of a waterbody
+            # replace that child with waterbody code.
+            new_conn[n] = []
+
+            for child in connections[n]:
+                if child in waterbodies:
+                    new_conn[n].append(waterbodies[child])
+                else:
+                    new_conn[n].append(child)
+        elif n in waterbodies:
+            wbody_code = waterbodies[n]
+
+            if wbody_code in new_conn:
+                continue
+
+            # find outflows of waterbody
+            outflows = set()
+            for tw in nhd_network.tailwaters(waterbody_nets[wbody_code]):
+                if tw in connections:
+                    outflows.update(connections[tw])
+            new_conn[wbody_code] = list(outflows)
+        else:
+            # copy to new network unchanged
+            new_conn[n] = connections[n]
+    return new_conn
```

Comment:
This block (lines 298-324) appears to be already in the main repository here:  
https://github.com/NOAA-OWP/t-route/blob/63e1578af5edf7086df7a4106314d5f31bfdbc71/src/python_framework_v02/troute/nhd_network.py#L404-L428